frontends/p4/substitutionVisitor.cpp

```diff
@@ -47,22 +47,25 @@ const IR::Node* TypeVariableSubstitutionVisitor::preorder(IR::TypeParameters *tp
     return tps;
 }
 
-const IR::Node* TypeVariableSubstitutionVisitor::preorder(IR::Type_Var* typeVariable) {
-    LOG1("Visiting " << dbp(getOriginal()));
-    const IR::Type* type = bindings->lookup(getOriginal<IR::Type_Var>());
-    if (type == nullptr)
-        return typeVariable;
-    LOG1("Replacing " << getOriginal() << " with " << type);
-    return type;
-}
-
-const IR::Node* TypeVariableSubstitutionVisitor::preorder(IR::Type_InfInt* typeVariable) {
-    LOG1("Visiting " << dbp(getOriginal()));
-    const IR::Type* type = bindings->lookup(getOriginal<IR::Type_InfInt>());
-    if (type == nullptr)
-        return typeVariable;
-    LOG1("Replacing " << getOriginal() << " with " << type);
-    return type;
+const IR::Node* TypeVariableSubstitutionVisitor::replacement(IR::ITypeVar* typeVariable) {
+    LOG3("Visiting " << dbp(getOriginal()));
+    const IR::ITypeVar* current = getOriginal<IR::ITypeVar>();
+    const IR::Type* replacement = nullptr;
+    while (true) {
+        // This will end because there should be no circular chain of variables pointing
+        // to each other.
+        const IR::Type* type = bindings->lookup(current);
+        if (type == nullptr)
+            break;
+        replacement = type;
+        if (!type->is<IR::ITypeVar>())
+            break;
+        current = type->to<IR::ITypeVar>();
+    }
```

Comment:
This loop could be written more concisely (and more clear?) as

    auto replacement = typeVariable->getNode();
    while(current) {
        if (auto type = bindings->lookup(current)) {
            replacement = type;
            current = type->to<IR::TypeVar>();
        } else {
            break; } }